src/drivers/uavcan_v1/ParamManager.hpp

```diff
@@ -0,0 +1,76 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2021 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file ParamManager.hpp
+ *
+ * Defines basic functionality of UAVCAN parameter management class
+ *
+ * @author Jacob Crabill <jacob@flyvoly.com>
+ */
+
+#pragma once
+
+#include <lib/parameters/param.h>
+
+#include <uavcan/_register/Name_1_0.h>
+#include <uavcan/_register/Value_1_0.h>
+
+typedef struct {
+	const char *uavcan_name;
+	const char *px4_name;
+	bool is_mutable {true};
+	bool is_persistent {true};
+} UavcanParamBinder;
```

Comment:
This is an okay approach but note that in the spec we agreed to make port-IDs be of type `natural16`, so this implementation is, strictly speaking, non-compliant. I think we could easily address this by adding a simple function here transforming a native parameter to the register value of the correct type, so that in your `UavcanParamBinder _uavcan_params[]` below you would go like:

```c++
const UavcanParamBinder _uavcan_params[8] {
    {
        "uavcan.pub.esc.0.id",
        "UCAN1_ESC_PUB",
        [](auto in, uavcan_register_Value_1_0& out){
            out.natural16.value.elements[0] = in;
            out.natural16.value.length = 1;
        },
    },
    ...
};
```

There is a much safer and cleaner approach based on metaprogramming but implementing that would require rewriting the parameter binder from scratch, not sure if anyone is interested. Otherwise, I can describe it here.

Reply:
> There is a much safer and cleaner approach based on metaprogramming but implementing that would require rewriting the parameter binder from scratch, not sure if anyone is interested. Otherwise, I can describe it here.

I would be very interested; I scratched my head for a while trying to think of how this should be done then got impatient and just implemented the simplest version I could to keep the momentum going.  Not sure if it would be something to include in this PR, but it could at least be a follow-up effort.

Reply:
Hi Pavel, I'm looking at this code above, but I'm a bit lost on how to implement this in PX4 where we've got C++17 but without the std library.

Thus we miss out on the std::function functionality, which makes it hard to store to the lambda, any thoughts on this?

Reply:
You can use an interface (abstract class) instead.